File: Math/CPP/matrix_cells_in_distance_order.cpp
#include <map>
#include <cmath>
#include <vector>

using namespace std;

class Solution {
public:
    vector<vector<int>> allCellsDistOrder(int rows, int cols, int rCenter, int cCenter) {
        map<int, vector<vector<int>>> hmap;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                vector<int> pos{i, j};
                hmap[dist(pos, rCenter, cCenter)].push_back(pos);
            }
        }

        vector<vector<int>> res;
        for (auto& [distance, cells] : hmap) {
            res.insert(res.end(), cells.begin(), cells.end());
        }

        return res;
    }

private:
    int dist(const vector<int>& pos, int rCenter, int cCenter) {
        return abs(pos[0] - rCenter) + abs(pos[1] - cCenter);
    }
};

File: Math/Python/matrix_cells_in_distance_order.py
from collections import defaultdict
from itertools import product

class Solution:
    def allCellsDistOrder(self, rows: int, cols: int, rCenter: int, cCenter: int) -> list[list[int]]:
        dist = lambda x: abs(x[0] - rCenter) + abs(x[1] - cCenter)
        hmap = defaultdict(list)
        mat = product(range(rows), range(cols))
        for coor in mat:
            hmap[dist(coor)].append(coor)
        res = []
        for i in range(max(hmap) + 1):
            res.extend(hmap[i])
        return res
